src/strategy.ts

```diff
@@ -53,66 +53,83 @@ export abstract class AbstractStrategy extends PassportStrategy {
       profile,
       loggedOut,
     }: {
-      profile?: Profile | null;
-      loggedOut?: boolean;
+      profile: Profile | null;
+      loggedOut: boolean;
     }) => {
       if (loggedOut) {
+        // Log out the current user no matter if we can verify the logged in user === logout requested user
         req.logout();
-        if (profile) {
+
+        // Check to see if we are logging out the user that is currently logged in to craft a proper IdP response
+        if (profile != null) {
           if (this._saml == null) {
             throw new Error("Can't get logout response URL without a SAML provider defined.");
           }
 
-          const RelayState =
-            (req.query && req.query.RelayState) || (req.body && req.body.RelayState);
-          return this._saml.getLogoutResponseUrl(profile, RelayState, options, redirectIfSuccess);
-        }
-        return this.pass();
-      }
-
-      const verified = (
-        err: Error | null,
-        user?: Record<string, unknown>,
-        info?: Record<string, unknown>
-      ) => {
-        if (err) {
-          return this.error(err);
+          const RelayState = req.query?.RelayState || req.body?.RelayState;
+
+          if (
+            req.user != null &&
+            req.user.ID === profile.ID &&
+            req.user.issuer === profile.issuer &&
+            req.user.nameID === profile.nameID &&
```

Comment:
If developer configures SamlStrategy e.g. like this (example taken from `passport-saml`'s `README.md`):
```javascript
passport.use(new SamlStrategy(
  {
    path: '/login/callback',
    entryPoint: 'https://openidp.feide.no/simplesaml/saml2/idp/SSOService.php',
    issuer: 'passport-saml',
    cert: 'fake cert', // cert must be provided
  },
  function(profile, done) {
    findByEmail(profile.email, function(err, user) {
      if (err) {
        return done(err);
      }
      return done(null, user);
    });
  })
);
```
and lets assume that `findByEmail` returns e.g.
```
{
  userId: 42
}
```
value of `req.user` and `req.session.passport.user` would be that.

I.e. there would not be any any of the attributes (like `nameID`) which are now used to match authenticated user to content of `LogoutRequest`.

I was expecting to see comparison (dunno how it could be implemented) of just `nameID` and `sessionIndex` (and maybe issuer)

Reply:
This is really the sticky part. Some of the types obfuscated what really needed to be compared here. I've fixed the types and tried again. This time we're leveraging the consumers code to make sure we're comparing the correct thing. Thank you very much for your sharp eye. Hopefully the changes I've made make sense, however, there is still a potential problem with the method I've chosen (see the TODO). I'll resolve this thread so we can start a new thread based on the new code.